### DISCUSSION TITLE
Duplicate Model Rendering for Different Models

### DISCUSSION THREAD

**Discussion by m6hdix** [RodrigoHamuy/react-three-map#122] [General]
I am experiencing a bug when rendering multiple 3D models using react-three-map and react-three-fiber. Despite having different URLs for the models, the same model is rendered across all canvases. This issue occurs when trying to load multiple models that are close to each other on the map.

Use the provided MapPolygons component.
Ensure there are at least three models with distinct URLs that are close to each other geographically.
Observe that the same model is rendered in all canvases, even though their URLs are different.
Expected Behavior:
Each canvas should render the 3D model specified by its unique URL.

Actual Behavior:
All canvases render the same model, ignoring the distinct URLs.

Relevant Code:

```import React, { memo, useEffect, useState, useRef, useCallback } from "react";
import { Layer, Source, useMap, Marker } from "react-map-gl";
import { useLoader } from "@react-three/fiber";
import { Canvas } from "react-three-map/maplibre";
import { FBXLoader } from "three/examples/jsm/loaders/FBXLoader";
import { HemisphereLight } from "three";
import { BORDER_COLORS } from "../../Services/Constants/BorderColors";
import { POLYGON_COLORS } from "../../Services/Constants/PolygonColors";
import useRequest from "../../Services/Hooks/useRequest";
import { ClipLoader } from "react-spinners";

const FBXModel = memo(({ url, rotation, setLoading, uniqueKey }) => {
  const fbx = useLoader(FBXLoader, url, (loader) => {
    loader.manager.onStart = () => setLoading(true);
    loader.manager.onLoad = () => setLoading(false);
    loader.manager.onError = () => setLoading(false);
  });

  const fbxRef = useRef();
  return (
    <group ref={fbxRef} rotation={rotation} scale={0.0097} key={uniqueKey}>
      <hemisphereLight
        args={["#ffffff", "#60666C"]}
        intensity={12}
        key={`${uniqueKey}-light`}
      />
      <primitive object={fbx} key={`${uniqueKey}-primitive`} />
    </group>
  );
});

const MapPolygons = () => {
  const map = useMap();
  const bounds = map.current.getBounds();
  const [features, setFeatures] = useState([]);
  const [buildingModels, setBuildingModels] = useState([]);
  const [zoom, setZoom] = useState(map.current.getZoom());
  const [isLoading, setIsLoading] = useState(false);

  const { Request } = useRequest();

  useEffect(() => {
    const handleViewportChange = () => {
      setZoom(map.current.getZoom());
    };

    map.current.on("zoomend", handleViewportChange);

    return () => {
      map.current.off("zoomend", handleViewportChange);
    };
  }, [map]);

  useEffect(() => {
    window.Echo.channel("feature-status").listen(
      ".feature-status-changed",
      (e) => {
        const data = [];
        for (const feature of features) {
          if (parseInt(feature.id) === parseInt(e.data.id)) {
            feature.rgb = e.data.rgb;
          }
          data.push(feature);
        }

        setFeatures(data);
      }
    );
  }, [features]);

  useEffect(() => {
    if (bounds.getSouthWest().lng && zoom >= 14) {
      const loadBuildings = zoom >= 15 ? "&load_buildings=1" : "";
      Request(
        `features?points[]=${bounds.getSouthWest().lng},${
          bounds.getSouthWest().lat
        }&points[]=${bounds.getSouthEast().lng},${
          bounds.getSouthEast().lat
        }&points[]=${bounds.getNorthWest().lng},${
          bounds.getNorthWest().lat
        }&points[]=${bounds.getNorthEast().lng},${
          bounds.getNorthEast().lat
        }${loadBuildings}`
      ).then((response) => {
        const newFeatures = [];
        const newBuildingModels = [];
        for (const feature of response?.data?.data) {
          newFeatures.push({
            id: feature?.geometry?.feature_id,
            rgb: feature?.properties?.rgb,
            coordinates: feature?.geometry?.coordinates.map((coordinate) => [
              parseFloat(coordinate.x),
              parseFloat(coordinate.y),
            ]),
          });

          if (feature.building_models) {
            for (const model of feature.building_models) {
              newBuildingModels.push(model);
            }
          }
        }

        setFeatures((prevFeatures) => [...prevFeatures, ...newFeatures]);
        setBuildingModels((prevModels) => [
          ...prevModels,
          ...newBuildingModels,
        ]);
      });
    }
  }, [bounds.getSouthWest().lng, zoom]);

  return (
    <>
      {zoom >= 14 && (
        <Source
          id="polygons"
          type="geojson"
          data={{
            type: "FeatureCollection",
            features: features.map((polygon) => ({
              type: "Feature",
              properties: {
                id: polygon.id,
                fill: POLYGON_COLORS[polygon.rgb],
                border: BORDER_COLORS[polygon.rgb],
              },
              geometry: {
                type: "Polygon",
                coordinates: [polygon.coordinates],
              },
            })),
          }}
        >
          <Layer
            id="polygon-fill-layer"
            type="fill"
            paint={{
              "fill-color": ["get", "fill"],
            }}
          />
          <Layer
            id="polygon-outline-layer"
            type="line"
            paint={{
              "line-color": ["get", "border"],
              "line-width": 3,
            }}
          />
        </Source>
      )}
      {zoom >= 15 &&
        buildingModels.length > 0 &&
        buildingModels.map((model) => {
          return (
            <>
              {isLoading && <ClipLoader color="#36d7b7" />}
              <Canvas
                latitude={parseFloat(model.building.position.split(", ")[0])}
                longitude={parseFloat(model.building.position.split(", ")[1])}
                key={`${model.id}-canvas`}
              >
                <FBXModel
                  url={model.file.url}
                  rotation={[0, 0, 0]}
                  setLoading={setIsLoading}
                  uniqueKey={`${model.id}-model`}
                />
              </Canvas>
            </>
          );
        })}
    </>
  );
};

export default MapPolygons;

```
#
The models have different URLs and unique keys.
This issue is impacting the accurate visualization of 3D models on the map.
Environment:

![image](https://github.com/user-attachments/assets/c2189cf0-65b0-478c-8bce-b8a4c52e4b35)

**Comment by m6hdix** [comment]
help sir

**Comment by m6hdix** [comment]
I would appreciate it if you could respond, I'm even willing to donate

**Comment by RodrigoHamuy** [comment]
Can you put this on a sandbox to be able to replicate? Here is a starting one you can fork https://codesandbox.io/p/sandbox/react-three-map-gettings-started-dhw34w

**Comment by m6hdix** [comment]
Hello RodrigoHamuy,

Thank you for reviewing the issue I raised. I have created a sandbox environment to demonstrate the problem. Please note that this is just for testing purposes and not intended for production.

Here is the link to the sandbox: [https://codesandbox.io/p/live/176d120c-fd63-4726-b612-7f48d5e13b30](url)

Please make sure to install and enable the CORS error extension, and for the environments to load correctly, zoom in to 15x.

I appreciate your time and look forward to your feedback.

Best regards
example video:


https://github.com/user-attachments/assets/40836d13-3be6-46b9-b274-40b3b36972ef

**Comment by RodrigoHamuy** [comment]
Do you have a minimum example without having to install custom extensions?

That sandbox is still a bit complicated for debugging.

But I can see you are creating a new `<Canvas>` per model. Instead, I would recommend to use [`<NearCoordinates>`](https://github.com/RodrigoHamuy/react-three-map?tab=readme-ov-file#nearcoordinates) instead.

**Reply by m6hdix** [reply]
I apologize for the delay in getting back to you.

We are aiming to load a specific 3D environment for each designated area and display it to the user accordingly. However, the main issue we are facing is that all the models are being rendered identically, even with delayed loading. While the addresses and the environments associated with each address are different, the environments are still being loaded in the same way, which is a bug.

Additionally, we want the environment to load only when the user approaches specific coordinates, rather than loading all three environments simultaneously.

Please take a look at the sandbox link below:
[https://codesandbox.io/p/devbox/2zq8ds?migrateFrom=4wvx49](https://github.com/RodrigoHamuy/react-three-map/issues/url)
Thank you for your attention to this matter.

Best regards,

**Reply by RodrigoHamuy** [reply]
@m6hdix I'm getting a "Devbox not found" error
![image](https://github.com/user-attachments/assets/77dee2d5-fd20-42b0-b5dd-70bf025efcb5)

**Reply by m6hdix** [reply]
[https://codesandbox.io/p/live/978b3b5d-4c86-4daf-a380-f5d09bb0621a](url)
hi bro ,new link

**Reply by RodrigoHamuy** [reply]
It's still 404, probably you have it set as a private sandbox?

**Reply by m6hdix** [reply]
[link](https://codesandbox.io/p/devbox/lgdflk?migrateFrom=4wvx49)
Hi ,I opened this new link in an incognito browser, no problem, thanks for your help

**Reply by RodrigoHamuy** [reply]
@m6hdix no idea why is not working, but as recommended before, you can use `NearCoordinates` instead of creating multiple canvas. This way is cheaper also. [link](https://codesandbox.io/p/devbox/react-three-map-discussion-122-fyhpnz)

**Reply by m6hdix** [reply]
@RodrigoHamuy  he models might be 200 kilometers apart, so it’s not a good idea to use a single canvas for all of them. I also tried using NearCoordinates, but it didn’t help.

**Reply by RodrigoHamuy** [reply]
@m6hdix then you can use [`Coordinates`](https://github.com/RodrigoHamuy/react-three-map?tab=readme-ov-file#coordinates). I updated the same codesandbox and things are working there. [link](https://codesandbox.io/p/devbox/react-three-map-discussion-122-fyhpnz). Normally you wouldn't want to have more than one `<Canvas>` in most cases.

```tsx
<Canvas latitude={51} longitude={0.2}>
  <Light />
  <Fbx path="./assets/3D-rgb-10230.fbx" />
  <Coordinates latitude={51.01} longitude={0.01}>
    <Light />
    <Fbx path="./assets/3D-rgb-10231.fbx" />
  </Coordinates>
  <Coordinates latitude={50.99} longitude={-0.01}>
    <Light />
    <Fbx path="./assets/3D-rgb-10235.fbx" />
  </Coordinates>
</Canvas>
```

**Comment by m6hdix** [comment]
[https://codesandbox.io/p/live/978b3b5d-4c86-4daf-a380-f5d09bb0621a](https://github.com/RodrigoHamuy/react-three-map/discussions/url)
hi bro ,new link
